### PR TITLE
PLAT-1260-1 Store user preferences in a JSON field

### PR DESCRIPTION
**Pull request by alexander-schmidt**
- `AGUser`: Replaced boolean preferences fields with a single JSON field.
  - Created `UserProfilePreferencesDiv[Test]`.
  - A DML migration to convert existing user preferences to JSON was deliberately omitted (deemed not worth the hassle because most users would stick with defaults anyway).
- Database migration test framework:
  - Added support for gaps between structure versions (e.g. `v24`, `v25`, `v30`) which may occur if there was no DB structure change between platform releases.
  - This PR contains the first DB structure change since platform v25.
  - That situation was unprecedented.

### CODE REVIEW

File: platform-core-module/src/main/java/com/softicar/platform/core/module/test/fixture/CoreModuleTestFixture.java [hunk 1]
```diff
@@ -49,17 +49,17 @@ public IModuleTestFixture<AGCoreModuleInstance> apply() {
 
 		this.viewUser = insertUser("View", "User")//
 			.setEmailAddress("view.user@example.com")
-			.setAutomaticallyCollapseFolders(false)
+			.updatePreferences(preferences -> preferences.setAutomaticallyCollapseFolders(false))
```
Comment (oliver-richers) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/test/fixture/CoreModuleTestFixture.java:52]: can we remove this?

File: platform-core-module/src/main/java/com/softicar/platform/core/module/user/AGUser.java [hunk 2]
```diff
@@ -247,4 +252,31 @@ public IDisplayString toDisplay() {
...
+
+	public UserPreferences getPreferences() {
+
+		UserPreferences preferences = null;
+		try {
+			preferences = new Gson().fromJson(getPreferencesJson(), UserPreferences.class);
+		} catch (Exception exception) {
+			DevNull.swallow(exception);
+			Log.ferror("Failed to retrieve preferences for user '%s'. Using defaults.", toDisplay());
+		}
```
Comment (oliver-richers) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/user/AGUser.java:264]: could be extracted to loadPreferences or something

File: platform-core-module/src/main/java/com/softicar/platform/core/module/user/UserPreferences.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
+package com.softicar.platform.core.module.user;
+
+public class UserPreferences {
+
+	public boolean automaticallyCollapseFolders = false;
+	public boolean recursivelyCollapseFolders = false;
```
Comment (oliver-richers) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/user/UserPreferences.java:6]: we should either make them private or remove the setters and getters, IMO

File: platform-core-module/src/main/java/com/softicar/platform/core/module/user/navigation/RecursivelyCollapseFoldersInput.java [hunk 4]
```diff
@@ -1,23 +1,22 @@
 package com.softicar.platform.core.module.user.navigation;
 
-import com.softicar.platform.core.module.user.AGUser;
+import com.softicar.platform.core.module.user.UserPreferences;
 import com.softicar.platform.emf.attribute.field.bool.EmfBooleanInput;
 
 public class RecursivelyCollapseFoldersInput extends EmfBooleanInput {
```
Comment (oliver-richers) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/user/navigation/RecursivelyCollapseFoldersInput.java:6]: should be deleted

File: platform-core-module/src/main/java/com/softicar/platform/core/module/user/profile/UserProfilePreferencesDiv.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+		public AutomaticallyCollapseFoldersCheckbox() {
+
+			super(preferences.automaticallyCollapseFolders);
+			addMarker(CoreTestMarker.USER_PREFERENCES_AUTOMATICALLY_COLLAPSE_FOLDERS_CHECKBOX);
+		}
+
+		@Override
+		public void refreshInputConstraints() {
+
+			preferences.automaticallyCollapseFolders = getValueOrThrow();
```
Comment (oliver-richers) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/user/profile/UserProfilePreferencesDiv.java:80]: this seems like a hack, why not use a change callback?

File: platform-core-module/src/main/java/com/softicar/platform/core/module/user/profile/UserProfilePreferencesDiv.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+		public void refreshInputConstraints() {
+
+			if (preferences.automaticallyCollapseFolders) {
+				setValue(true);
+				setDisabled(true);
+			} else {
+				setDisabled(false);
+			}
+
+			preferences.recursivelyCollapseFolders = getValueOrThrow();
```
Comment (oliver-richers) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/user/profile/UserProfilePreferencesDiv.java:102]: use change callback?

**Review by oliver-richers — CHANGES_REQUESTED** [review]

File: platform-core-module/src/main/java/com/softicar/platform/core/module/user/UserPreferences.java (hunk 3 above)
Comment (alexander-schmidt) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/user/UserPreferences.java:6]: ok

File: platform-core-module/src/main/java/com/softicar/platform/core/module/user/navigation/RecursivelyCollapseFoldersInput.java (hunk 4 above)
Comment (alexander-schmidt) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/user/navigation/RecursivelyCollapseFoldersInput.java:6]: Definitely. This was orphaned, and an oversight.

File: platform-core-module/src/main/java/com/softicar/platform/core/module/user/AGUser.java (hunk 2 above)
Comment (alexander-schmidt) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/user/AGUser.java]: ok

**Review by oliver-richers — APPROVED** [review]

**Pull request merged**